File: tools/verify.py
#!/usr/bin/env python3
import subprocess
import re
from pathlib import Path

from alchemy_gcc import CFLAGS, compiler_command

ROOT = Path(__file__).resolve().parents[1]
ROM_BASE = 0x08000000
def run(command):
    subprocess.run(command, cwd=ROOT, check=True,
                   stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def verify(source, rom, output_dir, details=False):
    address = int(source.stem, 16)
    symbol = f"Func_{address:08x}"
    assembly = output_dir / f"{source.stem}.s"
    obj = output_dir / f"{source.stem}.o"
    symbols_source = output_dir / f"{source.stem}.symbols.s"
    symbols_obj = output_dir / f"{source.stem}.symbols.o"
    elf = output_dir / f"{source.stem}.elf"
    binary = output_dir / f"{source.stem}.bin"
    run(compiler_command(
        *CFLAGS, "-S", "-o", assembly, source))
    run([
        "arm-none-eabi-as", "-mcpu=arm7tdmi", "-mthumb-interwork",
        "-o", str(obj), str(assembly),
    ])
    undefined = subprocess.check_output(
        ["arm-none-eabi-nm", "-u", str(obj)], text=True)
    names = []
    for line in undefined.splitlines():
        name = line.split()[-1]
        if not re.fullmatch(r"(Func|Data|Value)_[0-9a-f]{8}", name):
            raise ValueError(f"unsupported external symbol: {name}")
        names.append(name)
    symbols_source.write_text(
        ".syntax unified\n.thumb\n" +
        "".join(
            f".global {name}\n" +
            (".thumb_func\n" if name.startswith("Func_") else "") +
            f".set {name}, 0x{name.rsplit('_', 1)[1]}\n"
            for name in names
        )
    )
    run([
        "arm-none-eabi-as", "-mcpu=arm7tdmi", "-mthumb-interwork",
        "-o", str(symbols_obj), str(symbols_source),
    ])
    run([
        "arm-none-eabi-ld", f"-Ttext=0x{address:08x}", "-e", symbol,
        "-o", str(elf), str(obj), str(symbols_obj),
    ])
    run([
        "arm-none-eabi-objcopy", "-O", "binary", "-j", ".text",
        str(elf), str(binary),
    ])
    symbols = subprocess.check_output(
        ["arm-none-eabi-nm", "-S", str(elf)], text=True)
    row = next(line for line in symbols.splitlines()
               if line.endswith(" " + symbol))
    size = int(row.split()[1], 16)
    actual = binary.read_bytes()[:size]
    expected = rom[address - ROM_BASE:address - ROM_BASE + size]
    if details:
        return actual, expected, size
    return actual == expected, size


def main():
    rom = (ROOT / "baserom.gba").read_bytes()
    output_dir = ROOT / "out/verify"
    output_dir.mkdir(parents=True, exist_ok=True)
    sources = sorted((ROOT / "src").glob("*.c"))
    failures = []
    total = 0
    for source in sources:
        matched, size = verify(source, rom, output_dir)
        total += size
        if not matched:
            failures.append(source.name)
    print(f"verified={len(sources) - len(failures)} "
          f"failures={len(failures)} bytes={total}")
    if failures:
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: tools/verify.ts
#!/usr/bin/env bun
// Tool role: library; imported by tools/permute_m2c.ts, tools/permute_v1.ts.
import { mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { basename, dirname, extname, join } from "node:path";
import {
  cflagsForTargetSource,
  compilerCommandForTarget,
  externalSymbol,
  externalSymbolAssembly,
  type CompilerTarget,
} from "./alchemy_gcc.ts";
import { DEFAULT_TARGET, decompTarget, parseDecompTarget } from "./decomp_targets.ts";

export const ROOT = dirname(dirname(Bun.fileURLToPath(import.meta.url)));
export const ROM_BASE = 0x08000000;

function text(value: Uint8Array): string {
  return Buffer.from(value).toString("utf8");
}

function run(command: string[], quiet = true): string {
  const result = Bun.spawnSync(command, {
    cwd: ROOT,
    stdout: "pipe",
    stderr: "pipe",
  });
  if (result.exitCode !== 0) {
    const detail = (text(result.stderr) || text(result.stdout)).trim();
    if (quiet) throw new Error(`${basename(command[0])} failed`);
    throw new Error(`${basename(command[0])} failed: ${detail}`);
  }
  return text(result.stdout);
}

function stem(path: string): string {
  return basename(path, extname(path));
}

export function verify(
  source: string,
  rom: Uint8Array,
  outputDir: string,
  details: true,
  compiler?: CompilerTarget,
): [Buffer, Buffer, number];
export function verify(
  source: string,
  rom: Uint8Array,
  outputDir: string,
  details?: false,
  compiler?: CompilerTarget,
): [boolean, number];
export function verify(
  source: string,
  rom: Uint8Array,
  outputDir: string,
  details = false,
  compiler: CompilerTarget = "gs1",
): [Buffer, Buffer, number] | [boolean, number] {
  const name = stem(source);
  const address = Number.parseInt(name, 16);
  const symbol = `Func_${address.toString(16).padStart(8, "0")}`;
  const assembly = join(outputDir, `${name}.s`);
  const object = join(outputDir, `${name}.o`);
  const symbolsSource = join(outputDir, `${name}.symbols.s`);
  const symbolsObject = join(outputDir, `${name}.symbols.o`);
  const elf = join(outputDir, `${name}.elf`);
  const binary = join(outputDir, `${name}.bin`);
  run(compilerCommandForTarget(
    compiler,
    ...cflagsForTargetSource(compiler, source),
    "-S", "-o", assembly, source,
  ));
  run([
    "arm-none-eabi-as", "-mcpu=arm7tdmi", "-mthumb-interwork",
    "-o", object, assembly,
  ]);
  const names: string[] = [];
  for (const line of run(["arm-none-eabi-nm", "-u", object]).split(/\r?\n/).filter(Boolean)) {
    const external = line.trim().split(/\s+/).at(-1)!;
    if (externalSymbol(external) === null) throw new Error(`unsupported external symbol: ${external}`);
    names.push(external);
  }
  writeFileSync(symbolsSource, ".syntax unified\n.thumb\n" + names.map(externalSymbolAssembly).join(""));
  run([
    "arm-none-eabi-as", "-mcpu=arm7tdmi", "-mthumb-interwork",
    "-o", symbolsObject, symbolsSource,
  ]);
  run([
    "arm-none-eabi-ld", `-Ttext=0x${address.toString(16).padStart(8, "0")}`,
    "-e", symbol, "-o", elf, object, symbolsObject,
  ]);
  run(["arm-none-eabi-objcopy", "-O", "binary", "-j", ".text", elf, binary]);
  const symbols = run(["arm-none-eabi-nm", "-S", elf]);
  const row = symbols.split(/\r?\n/).find((line) => line.endsWith(` ${symbol}`));
  if (row === undefined) throw new Error(`missing linked symbol: ${symbol}`);
  const size = Number.parseInt(row.trim().split(/\s+/)[1], 16);
  const actual = readFileSync(binary).subarray(0, size);
  const expected = Buffer.from(rom).subarray(address - ROM_BASE, address - ROM_BASE + size);
  if (details) return [actual, expected, size];
  return [actual.equals(expected), size];
}

function main(): void {
  const arguments_ = Bun.argv.slice(2);
  if (arguments_.some((argument) => argument === "-h" || argument === "--help")) {
    console.log("usage: verify.ts [-h] [--target gs1-en|gs2-en]");
    return;
  }
  let targetId = DEFAULT_TARGET;
  for (let index = 0; index < arguments_.length; index++) {
    const argument = arguments_[index];
    if (argument === "--target") {
      const value = arguments_[++index];
      if (value === undefined) throw new Error("--target requires a value");
      targetId = parseDecompTarget(value);
    } else if (argument.startsWith("--target=")) {
      targetId = parseDecompTarget(argument.slice(9));
    } else {
      throw new Error(`unrecognized argument: ${argument}`);
    }
  }
  const target = decompTarget(targetId);
  const rom = readFileSync(join(ROOT, target.rom));
  if (rom.length !== target.romSize) {
    throw new Error(`${target.id} ROM must contain exactly ${target.romSize} bytes`);
  }
  const outputDir = join(ROOT, target.outputDir, "verify");
  mkdirSync(outputDir, { recursive: true });
  const sourceDirectory = join(ROOT, target.sourceDir);
  const sources = readdirSync(sourceDirectory, { withFileTypes: true })
    .filter((entry) => entry.isFile() && entry.name.endsWith(".c"))
    .map((entry) => join(sourceDirectory, entry.name)).sort();
  const failures: string[] = [];
  let total = 0;
  for (const source of sources) {
    const [matched, size] = verify(source, rom, outputDir, false, target.compiler);
    total += size;
    if (!matched) failures.push(basename(source));
  }
  console.log(`verified=${sources.length - failures.length} failures=${failures.length} bytes=${total}`);
  if (failures.length !== 0) process.exitCode = 1;
}

if (import.meta.main) main();
